src/transforms/groupby.js

```diff
@@ -82,43 +90,68 @@ exports.transform = function(data, state) {
 
     var newData = [];
 
-    data.forEach(function(trace) {
-        newData = newData.concat(transformOne(trace, state));
+    data.forEach(function(trace, i) {
+
+        newData = newData.concat(transformOne(trace, state, state.attributeSets[i]));
     });
 
     return newData;
 };
 
-function transformOne(trace, state) {
+function initializeArray(newTrace, a) {
+    Lib.nestedProperty(newTrace, a).set([]);
+}
+
+function pasteArray(newTrace, trace, j, a) {
+    Lib.nestedProperty(newTrace, a).set(
+        Lib.nestedProperty(newTrace, a).get().concat([
+            Lib.nestedProperty(trace, a).get()[j]
+        ])
+    );
+}
+
+function transformOne(trace, state, attributeSet) {
+
     var opts = state.transform;
-    var groups = opts.groups;
+    var groups = trace.transforms[state.transformIndex].groups;
 
     var groupNames = groups.filter(function(g, i, self) {
         return self.indexOf(g) === i;
     });
 
+    if(!(Array.isArray(groups)) || groups.length === 0) {
+        return trace;
+    }
+
     var newData = new Array(groupNames.length);
-    var len = Math.min(trace.x.length, trace.y.length, groups.length);
+    var len = groups.length;
 
+    var style = opts.style || {};
+
+    var arrayAttributes = attributeSet
+        .filter(function(array) {return Array.isArray(Lib.nestedProperty(trace, array).get());});
+
+    // fixme the O(n**3) complexity
```

Comment:
I'll remove the comment and iterate on performance either separately or as needed (as you see fit)